File: ios/Providers/DatabaseProvider.swift
//
//  DatabaseProvider.swift
//  Glimpse
//
//  Created by Nitin Issac Joy on 10/07/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import RealmSwift

class DatabaseProvider {
  
  static func ClearAllTrips() {
    let db = try! Realm()
    let dbObjects = db.objects(TripModal.self)
    for object in dbObjects {
      try! db.write {
        db.delete(object)
      }
    }
    let dbObjects2 = db.objects(StepModal.self)
    for object in dbObjects2 {
      try! db.write {
        db.delete(object)
      }
    }
  }
  
  static func UpdateDBWithTrips(trips: [TripModal]) {
    let db = try! Realm();
    
    for trip in trips{
      let dbResult = db.objects(TripModal.self).filter{ $0.tripId == trip.tripId }
      if let _trip = dbResult.first {
        try! db.write {
          _trip.CopyConstructor(trip: trip)
        }
      } else {
        try! db.write {
          db.add(trip)
        }
      }
    }
  }
  
  static func UpdateDBWithSteps(steps: [StepModal]) {
    let db = try! Realm()
    for step in steps {
      let dbObjects = db.objects(StepModal.self).filter{ $0.tripId == step.tripId && $0.stepId == step.stepId }
      if let obj = dbObjects.first {
        try! db.write {
          obj.CopyConstructor(step: step);
        }
      } else {
        try! db.write {
          db.add(step)
        }
      }
    }
  }
  
  static func UpdateDBWithHomesForDataClustering(homes: [HomesForDataClusteringModal]) {
    let db = try! Realm()
    let dbresult = db.objects(HomesForDataClusteringModal.self)
    try! db.write {
      db.delete(dbresult)
    }
    
    let dbData: List<HomesForDataClusteringModal> = List<HomesForDataClusteringModal>()
    for home in homes {
      dbData.append(home)
    }
    try! db.write {
      db.add(dbData)
    }
  }
  
  static func GetHomesForDataClustering() -> [HomesForDataClusteringModal] {
    let db = try! Realm()
    let dbresult = db.objects(HomesForDataClusteringModal.self)
    
    var returnArray: [HomesForDataClusteringModal] = []
    for res in dbresult {
      let home = HomesForDataClusteringModal()
      home.CopyConstructor(home: res)
      returnArray.append(home)
    }
    
    return returnArray
  }
}

File: android/app/src/main/java/com/beerwithai/glimpse/Engine/Providers/DatabaseProvider.java
package com.beerwithai.glimpse.Engine.Providers;

import com.beerwithai.glimpse.Engine.Modals.*;

import java.util.ArrayList;

import io.realm.Realm;
import io.realm.RealmResults;

public class DatabaseProvider {

    public static void ClearAllTrips() {
        Realm db = Realm.getDefaultInstance();

        try {
            db.beginTransaction();
            RealmResults<TripModal> dbObjects = db.where(TripModal.class).findAll();
            dbObjects.deleteAllFromRealm();
            RealmResults<StepModal> dbObjects2 = db.where(StepModal.class).findAll();
            dbObjects2.deleteAllFromRealm();
            db.commitTransaction();
        } finally {
            db.close();
        }
    }

    public static void UpdateDBWithTrips(ArrayList<TripModal> trips) {
        Realm db = Realm.getDefaultInstance();

        try {
            for (int i = 0; i < trips.size(); i++) {
                TripModal dbObject = db.where(TripModal.class).equalTo("tripId", trips.get(i).tripId).findFirst();
                if (dbObject != null) { //TODO: Check if this is the case
                    db.beginTransaction();
                    dbObject.CopyConstructor(trips.get(i));
                    db.commitTransaction();
                } else {
                    db.beginTransaction();
                    db.copyToRealm(trips.get(i));
                    db.commitTransaction();
                }
            }
        } finally {
            db.close();
        }

    }

    public static void UpdateDBWithSteps(ArrayList<StepModal> steps) {
        Realm db = Realm.getDefaultInstance();

        try {
            for (int i = 0; i < steps.size(); i++) {
                StepModal dbObject = db.where(StepModal.class).equalTo("tripId", steps.get(i).tripId).equalTo("stepId", steps.get(i).stepId).findFirst();
                if (dbObject != null) { //TODO: Check if this is the case
                    db.beginTransaction();
                    dbObject.CopyConstructor(steps.get(i));
                    db.commitTransaction();
                } else {
                    db.beginTransaction();
                    db.copyToRealm(steps.get(i));
                    db.commitTransaction();
                }
            }
        } catch(Exception e) {

        }
    }

    public static void UpdateDBWithHomesForDataClustering(ArrayList<HomesForDataClusteringModal> homes) {
        Realm db = Realm.getDefaultInstance();

        try {
            db.beginTransaction();
            db.where(HomesForDataClusteringModal.class).findAll().deleteAllFromRealm();
            for(int i = 0; i < homes.size(); i++) {
                db.copyToRealm(homes.get(i));
            }
            db.commitTransaction();
        } catch (Exception e) {

        }
    }

    public static HomesForDataClusteringModal[] GetHomesForDataClustering() {
        Realm db = Realm.getDefaultInstance();

        try {
            RealmResults<HomesForDataClusteringModal> dbResults = db.where(HomesForDataClusteringModal.class).findAll();
            HomesForDataClusteringModal[] results = new HomesForDataClusteringModal[dbResults.size()];
            for(int i = 0; i < dbResults.size(); i++) {
                results[i] = dbResults.get(i);
            }
            return results;
        } catch (Exception e) {
            return null;
        }
    }


}
